README.rst
tensorboard_logger
==================


.. image:: https://img.shields.io/pypi/v/tensorboard_logger.svg
        :target: https://pypi.python.org/pypi/tensorboard_logger

.. image:: https://img.shields.io/travis/TeamHG-Memex/tensorboard_logger.svg
        :target: https://travis-ci.org/TeamHG-Memex/tensorboard_logger


Log TensorBoard events without TensorFlow
-----------------------------------------

`TensorBoard <https://www.tensorflow.org/how_tos/summaries_and_tensorboard/>`_
is a visualization tool (not this project, it's a part of
`TensorFlow <https://www.tensorflow.org>`_ framework)
that makes it easy to check training progress, compare between
different runs, and has lots of other cool features.

.. image:: tensorboard_example.png

``tensorboard_logger`` library allows to write TensorBoard events without TensorFlow::

    from tensorboard_logger import configure, log_value

    configure("runs/run-1234")

    for step in range(1000):
        v1, v2 = do_stuff()
        log_value('v1', v1, step)
        log_value('v2', v2, step)


Note: if you are already using TensorFlow in your project,
you probably don't need this library.


Installation
------------

TensorFlow is required only for viewing logged events: please check installation guide
on the `official site <https://www.tensorflow.org/get_started/os_setup.html#download-and-setup>`_
(you probably want a CPU-only version).

``tensorboard_logger`` can be installed with pip::

    pip install tensorboard_logger


Usage
-----

You can either use default logger with ``tensorboard_logger.configure``
and ``tensorboard_logger.log_value`` functions, or use ``tensorboard_logger.Logger`` class.

This library can be used to log numerical values of some variables in TensorBoard format, so you can
use TensorBoard to visualize how they changed, and compare same variables between different runs.
Log file is written into a directory, so you need a separate directory for each run
(you can place other logs or output files you use in the same directory).
Directories from different runs you wish to compare should have the same parent
(there can be other files or directories with the same parent, TensorBoard will figure out
which directories contain logs).

Apart from variable names and their values, another important thing is the **step**: this must
be an integer that represents some increasing step - it can be a step in training or some
other number. The values are ordered by step in TensorBoard, although you can view them
ordered by time or relative step too.

A simple usage example::

    from tensorboard_logger import configure, log_value

    configure("runs/run-1234", flush_secs=5)

    for step in range(1000):
        v1, v2 = do_stuff()
        log_value('v1', v1, step)
        log_value('v2', v2, step)


You can start TensorBoard right away::

    tensorboard --logdir runs

And go check the metrics to TensorBoard UI at http://localhost:6006
(note that it binds to 0.0.0.0 by default).
Metrics are refreshed on switch to browser tab, and there is also a refresh button
at the top right.

Runtime overhead rather large, about 0.1 - 0.2 ms for a single value logged
(so about 5,000 - 10,000 operations per second).


API
---

``tensorboard_logger.configure(logdir, flush_secs=2)``

Configure logging: a file will be written to ``logdir``, and flushed every ``flush_secs``.
**NOTE:** right now file is flushed after each event written.

``tensorboard_logger.log_value(name, value, step=None)``

Log new ``value`` for given ``name`` on given ``step``.
``value`` should be a real number (it will be converted to float),
and ``name`` should be a string (it will be converted to a valid
TensorFlow summary name). ``step`` should be an non-negative integer,
and is used for visualization: you can log several different
variables on one step, but should not log different values
of the same variable on the same step (this is not checked).
You can also omit step entirely.

``tensorboard_logger.Logger``

A class for writing logs in a directory.
Use it if default logger used in two above functions is not enough for you
(e.g. you want to log into several different directories, or don't like
global variables).
Constructor has the same signature as ``tensorboard_logger.configure``,
and it has a single ``log_value`` method with the same signature as
``tensorboard_logger.log_value``.


Development
-----------

Compiling python protobuf files::

    protoc --python_out . tensorboard_logger/tf_protobuf/summary.proto
    protoc --python_out . tensorboard_logger/tf_protobuf/event.proto


License
-------

`MIT license`_

.. _MIT license: LICENSE

----

.. image:: https://hyperiongray.s3.amazonaws.com/define-hg.svg
	:target: https://hyperiongray.com/?pk_campaign=github&pk_kwd=tensorboard_logger
	:alt: define hyperiongray


HISTORY.rst
History
=======

0.1.0 (2018-02-08)
------------------

* Ability to log histograms and images (#18)


0.0.4 (2017-07-31)
------------------

* Fix working with TensorFlow installed (#11)


0.0.3 (2016-11-17)
------------------

* Fix PyPI packaging


0.0.2 (2016-11-03)
------------------

* TensorFlow is no longer required (#1)
* Travis build is fixed (#2)


0.0.1 (2016-11-01)
------------------

* First release on PyPI.


tests/test_tensorboard_logger.py
# -*- coding: utf-8 -*-
import time
import os
import glob
import numpy as np

from tensorboard_logger import Logger, configure, log_value
from tensorboard_logger.tensorboard_logger import make_valid_tf_name


def test_smoke_default(tmpdir):
    configure(str(tmpdir), flush_secs=0.1)
    for step in range(10):
        log_value('v1', step * 1.5, step)
        log_value('v2', step ** 1.5 - 2)
    time.sleep(0.5)
    tf_log, = tmpdir.listdir()
    assert tf_log.basename.startswith('events.out.tfevents.')


def test_smoke_logger(tmpdir):
    logger = Logger(str(tmpdir), flush_secs=0.1)
    for step in range(10):
        logger.log_value('v1', step * 1.5, step)
        logger.log_value('v2', step ** 1.5 - 2)
    time.sleep(0.5)
    tf_log, = tmpdir.listdir()
    assert tf_log.basename.startswith('events.out.tfevents.')


def test_serialization(tmpdir):
    logger = Logger(str(tmpdir), flush_secs=0.1, dummy_time=256.5)
    logger.log_value('v/1', 1.5, 1)
    logger.log_value('v/22', 16.0, 2)
    time.sleep(0.5)
    tf_log, = tmpdir.listdir()
    assert tf_log.read_binary() == (
        # step = 0, initial record
        b'\x18\x00\x00\x00\x00\x00\x00\x00\xa3\x7fK"\t\x00\x00\x00\x00\x00\x08p@\x1a\rbrain.Event:2\xbc\x98!+'
        # v/1
        b'\x19\x00\x00\x00\x00\x00\x00\x00\x8b\xf1\x08(\t\x00\x00\x00\x00\x00\x08p@\x10\x01*\x0c\n\n\n\x03v/1\x15\x00\x00\xc0?,\xec\xc0\x87'
        # v/22
        b'\x1a\x00\x00\x00\x00\x00\x00\x00\x12\x9b\xd8-\t\x00\x00\x00\x00\x00\x08p@\x10\x02*\r\n\x0b\n\x04v/22\x15\x00\x00\x80A\x8f\xa3\xb6\x88'
    )


def test_dummy():
    logger = Logger(None, is_dummy=True)
    for step in range(3):
        logger.log_value('A v/1', step, step)
        logger.log_value('A v/2', step * 2, step)
    assert dict(logger.dummy_log) == {
        'A_v/1': [(0, 0), (1, 1), (2, 2)],
        'A_v/2': [(0, 0), (1, 2), (2, 4)],
    }


def test_make_valid_tf_name():
    mvn = make_valid_tf_name
    assert mvn('This/is/valid') == 'This/is/valid'
    assert mvn('0-This/is/valid') == '0-This/is/valid'
    assert mvn('.This/is/valid') == '.This/is/valid'
    assert mvn(' This/is invalid') == '._This/is_invalid'
    assert mvn('-This-is-invalid') == '.-This-is-invalid'


def test_unique():
    logger = Logger(None, is_dummy=True)
    for step in range(1, 3):
        # names that normalize to the same valid name
        logger.log_value('A v/1', step, step)
        logger.log_value('A\tv/1', step * 2, step)
        logger.log_value('A  v/1', step * 3, step)
    assert dict(logger.dummy_log) == {
        'A_v/1':   [(1, 1), (2, 2)],
        'A_v/1/1': [(1, 2), (2, 4)],
        'A_v/1/2': [(1, 3), (2, 6)],
    }


def test_dummy_histo():
    logger = Logger(None, is_dummy=True)
    bins = [0, 1, 2, 3]
    logger.log_histogram('key', (bins, [0.0, 1.0, 2.0]), step=1)
    logger.log_histogram('key', (bins, [1.0, 1.5, 2.5]), step=2)
    logger.log_histogram('key', (bins, [0.0, 1.0, 2.0]), step=3)

    assert dict(logger.dummy_log) == {
        'key': [(1, (bins, [0.0, 1.0, 2.0])),
                (2, (bins, [1.0, 1.5, 2.5])),
                (3, (bins, [0.0, 1.0, 2.0]))]}


def test_real_histo_tuple(tmpdir):
    """
    from tests.test_tensorboard_logger import *
    import ubelt as ub
    ub.delete(ub.ensure_app_cache_dir('tf_logger'))
    tmpdir = ub.ensure_app_cache_dir('tf_logger/runs/run1')
    """
    logger = Logger(str(tmpdir), flush_secs=0.1)
    bins = [-.5, .5, 1.5, 2.5]
    logger.log_histogram('hist1', (bins, [0.0, 1.0, 2.0]), step=1)
    logger.log_histogram('hist1', (bins, [1.0, 1.5, 2.5]), step=2)
    logger.log_histogram('hist1', (bins, [0.0, 1.0, 2.0]), step=3)
    tf_log, = glob.glob(str(tmpdir) + '/*')
    assert os.path.basename(tf_log).startswith('events.out.tfevents.')


def test_real_histo_data(tmpdir):
    logger = Logger(str(tmpdir), flush_secs=0.1)
    logger.log_histogram('hist2', [1, 7, 6, 9, 8, 1, 4, 5, 3, 7], step=1)
    logger.log_histogram('hist2', [5, 3, 2, 0, 8, 5, 7, 7, 7, 2], step=2)
    logger.log_histogram('hist2', [1, 2, 2, 1, 5, 1, 8, 4, 4, 1], step=3)
    tf_log, = glob.glob(str(tmpdir) + '/*')
    assert os.path.basename(tf_log).startswith('events.out.tfevents.')


def test_dummy_images():
    logger = Logger(None, is_dummy=True)
    img = np.random.rand(10, 10)
    images = [img, img]
    logger.log_images('key', images, step=1)
    logger.log_images('key', images, step=2)
    logger.log_images('key', images, step=3)

    assert dict(logger.dummy_log) == {
        'key': [(1, images),
                (2, images),
                (3, images)]}


def test_real_image_data(tmpdir):
    logger = Logger(str(tmpdir), flush_secs=0.1)
    img = np.random.rand(10, 10)
    images = [img, img]
    logger.log_images('key', images, step=1)
    logger.log_images('key', images, step=2)
    logger.log_images('key', images, step=3)
    tf_log, = glob.glob(str(tmpdir) + '/*')
    assert os.path.basename(tf_log).startswith('events.out.tfevents.')


travis_pypi_setup.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Update encrypted deploy password in Travis config file
"""


from __future__ import print_function
import base64
import json
import os
from getpass import getpass
import yaml
from cryptography.hazmat.primitives.serialization import load_pem_public_key
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15


try:
    from urllib import urlopen
except:
    from urllib.request import urlopen


GITHUB_REPO = 'TeamHG-Memex/tensorboard_logger'
TRAVIS_CONFIG_FILE = os.path.join(
    os.path.dirname(os.path.abspath(__file__)), '.travis.yml')


def load_key(pubkey):
    """Load public RSA key, with work-around for keys using
    incorrect header/footer format.

    Read more about RSA encryption with cryptography:
    https://cryptography.io/latest/hazmat/primitives/asymmetric/rsa/
    """
    try:
        return load_pem_public_key(pubkey.encode(), default_backend())
    except ValueError:
        # workaround for https://github.com/travis-ci/travis-api/issues/196
        pubkey = pubkey.replace('BEGIN RSA', 'BEGIN').replace('END RSA', 'END')
        return load_pem_public_key(pubkey.encode(), default_backend())


def encrypt(pubkey, password):
    """Encrypt password using given RSA public key and encode it with base64.

    The encrypted password can only be decrypted by someone with the
    private key (in this case, only Travis).
    """
    key = load_key(pubkey)
    encrypted_password = key.encrypt(password, PKCS1v15())
    return base64.b64encode(encrypted_password)


def fetch_public_key(repo):
    """Download RSA public key Travis will use for this repo.

    Travis API docs: http://docs.travis-ci.com/api/#repository-keys
    """
    keyurl = 'https://api.travis-ci.org/repos/{0}/key'.format(repo)
    data = json.loads(urlopen(keyurl).read().decode())
    if 'key' not in data:
        errmsg = "Could not find public key for repo: {}.\n".format(repo)
        errmsg += "Have you already added your GitHub repo to Travis?"
        raise ValueError(errmsg)
    return data['key']


def prepend_line(filepath, line):
    """Rewrite a file adding a line to its beginning.
    """
    with open(filepath) as f:
        lines = f.readlines()

    lines.insert(0, line)

    with open(filepath, 'w') as f:
        f.writelines(lines)


def load_yaml_config(filepath):
    with open(filepath) as f:
        return yaml.load(f)


def save_yaml_config(filepath, config):
    with open(filepath, 'w') as f:
        yaml.dump(config, f, default_flow_style=False)


def update_travis_deploy_password(encrypted_password):
    """Update the deploy section of the .travis.yml file
    to use the given encrypted password.
    """
    config = load_yaml_config(TRAVIS_CONFIG_FILE)

    config['deploy']['password'] = dict(secure=encrypted_password)

    save_yaml_config(TRAVIS_CONFIG_FILE, config)

    line = ('# This file was autogenerated and will overwrite'
            ' each time you run travis_pypi_setup.py\n')
    prepend_line(TRAVIS_CONFIG_FILE, line)


def main(args):
    public_key = fetch_public_key(args.repo)
    password = args.password or getpass('PyPI password: ')
    update_travis_deploy_password(encrypt(public_key, password.encode()))
    print("Wrote encrypted password to .travis.yml -- you're ready to deploy")


if '__main__' == __name__:
    import argparse
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--repo', default=GITHUB_REPO,
                        help='GitHub repo (default: %s)' % GITHUB_REPO)
    parser.add_argument('--password',
                        help='PyPI password (will prompt if not provided)')

    args = parser.parse_args()
    main(args)


tensorboard_logger/tensorboard_logger.py
# -*- coding: utf-8 -*-
from collections import defaultdict
import os
import re
import socket
import struct
import time
import numpy as np

import six

import scipy.misc
try:
    from StringIO import StringIO  # Python 2.7
except ImportError:
    from io import BytesIO         # Python 3.x

try:
    from tensorflow.core.util import event_pb2
    from tensorflow.core.framework import summary_pb2
except ImportError:
    from .tf_protobuf import summary_pb2, event_pb2
from .crc32c import crc32c


__all__ = ['Logger', 'configure', 'log_value', 'log_histogram', 'log_images']


_VALID_OP_NAME_START = re.compile('^[A-Za-z0-9.]')
_VALID_OP_NAME_PART = re.compile('[A-Za-z0-9_.\\-/]+')


class Logger(object):
    def __init__(self, logdir, flush_secs=2, is_dummy=False, dummy_time=None):
        self._name_to_tf_name = {}
        self._tf_names = set()
        self.is_dummy = is_dummy
        self.logdir = logdir
        self.flush_secs = flush_secs  # TODO
        self._writer = None
        self._dummy_time = dummy_time
        if is_dummy:
            self.dummy_log = defaultdict(list)
        else:
            if not os.path.exists(self.logdir):
                os.makedirs(self.logdir)
            hostname = socket.gethostname()
            filename = os.path.join(
                self.logdir, 'events.out.tfevents.{}.{}'.format(
                    int(self._time()), hostname))
            self._writer = open(filename, 'wb')
            self._write_event(event_pb2.Event(
                wall_time=self._time(), step=0, file_version='brain.Event:2'))

    def _ensure_tf_name(self, name):
        if not isinstance(name, six.string_types):
            raise TypeError('"name" should be a string, got {}'
                            .format(type(name)))
        try:
            tf_name = self._name_to_tf_name[name]
        except KeyError:
            tf_name = self._make_tf_name(name)
            self._name_to_tf_name[name] = tf_name
        return tf_name

    def _check_step(self, step):
        if step is not None and not isinstance(step, six.integer_types):
            raise TypeError('"step" should be an integer, got {}'
                            .format(type(step)))

    def log_value(self, name, value, step=None):
        """Log new value for given name on given step.

        Args:
            name (str): name of the variable (it will be converted to a valid
                tensorflow summary name).
            value (float): this is a real number to be logged as a scalar.
            step (int): non-negative integer used for visualization: you can
                log several different variables on one step, but should not log
                different values of the same variable on the same step (this is
                not checked).
        """
        if isinstance(value, six.string_types):
            raise TypeError('"value" should be a number, got {}'
                            .format(type(value)))
        value = float(value)

        self._check_step(step)
        tf_name = self._ensure_tf_name(name)

        summary = self._scalar_summary(tf_name, value, step)
        self._log_summary(tf_name, summary, value, step=step)

    def log_histogram(self, name, value, step=None):
        """Log a histogram for given name on given step.

        Args:
            name (str): name of the variable (it will be converted to a valid
                tensorflow summary name).
            value (tuple or list): either list of numbers
                to be summarized as a histogram, or a tuple of bin_edges and
                bincounts that directly define a histogram.
            step (int): non-negative integer used for visualization
        """
        if isinstance(value, six.string_types):
            raise TypeError('"value" should be a number, got {}'
                            .format(type(value)))

        self._check_step(step)
        tf_name = self._ensure_tf_name(name)

        summary = self._histogram_summary(tf_name, value, step=step)
        self._log_summary(tf_name, summary, value, step=step)

    def log_images(self, name, images, step=None):
        """Log new images for given name on given step.

        Args:
            name (str): name of the variable (it will be converted to a valid
                tensorflow summary name).
            images (list): list of images to visualize
            step (int): non-negative integer used for visualization
        """
        if isinstance(images, six.string_types):
            raise TypeError('"images" should be a list of ndarrays, got {}'
                            .format(type(images)))

        self._check_step(step)
        tf_name = self._ensure_tf_name(name)

        summary = self._image_summary(tf_name, images, step=step)
        self._log_summary(tf_name, summary, images, step=step)

    def _image_summary(self, tf_name, images, step=None):
        """
        Log a list of images.

        References:
            https://github.com/yunjey/pytorch-tutorial/blob/master/tutorials/04-utils/tensorboard/logger.py#L22

        Example:
            >>> tf_name = 'foo'
            >>> value = ([0, 1, 2, 3, 4, 5], [1, 20, 10, 22, 11])
            >>> self = Logger(None, is_dummy=True)
            >>> images = [np.random.rand(10, 10), np.random.rand(10, 10)]
            >>> summary = self._image_summary(tf_name, images, step=None)
            >>> assert len(summary.value) == 2
            >>> assert summary.value[0].image.width == 10
        """
        img_summaries = []
        for i, img in enumerate(images):
            # Write the image to a string
            try:
                s = StringIO()
            except:
                s = BytesIO()
            scipy.misc.toimage(img).save(s, format="png")

            # Create an Image object
            img_sum = summary_pb2.Summary.Image(
                encoded_image_string=s.getvalue(),
                height=img.shape[0],
                width=img.shape[1]
            )
            # Create a Summary value
            img_value = summary_pb2.Summary.Value(tag='{}/{}'.format(tf_name, i),
                                                  image=img_sum)
            img_summaries.append(img_value)
            summary = summary_pb2.Summary()
            summary.value.add(tag=tf_name, image=img_sum)

        summary = summary_pb2.Summary(value=img_summaries)
        return summary

    def _histogram_summary(self, tf_name, value, step=None):
        """
        Args:
            tf_name (str): name of tensorflow variable
            value (tuple or list): either a tuple of bin_edges and bincounts or
                a list of values to summarize in a histogram.

        References:
            https://github.com/yunjey/pytorch-tutorial/blob/master/tutorials/04-utils/tensorboard/logger.py#L45

        Example:
            >>> tf_name = 'foo'
            >>> value = ([0, 1, 2, 3, 4, 5], [1, 20, 10, 22, 11])
            >>> self = Logger(None, is_dummy=True)
            >>> summary = self._histogram_summary(tf_name, value, step=None)
            >>> assert summary.value[0].histo.max == 5

        Example:
            >>> tf_name = 'foo'
            >>> value = [0.72,  0.18,  0.34,  0.66,  0.11,  0.70,  0.23]
            >>> self = Logger(None, is_dummy=True)
            >>> summary = self._histogram_summary(tf_name, value, step=None)
            >>> assert summary.value[0].histo.num == 7.0
        """
        if isinstance(value, tuple):
            bin_edges, bincounts = value
            assert len(bin_edges) == len(bincounts) + 1, (
                'must have one more edge than count')
            hist = summary_pb2.HistogramProto()
            hist.min = float(min(bin_edges))
            hist.max = float(max(bin_edges))
        else:
            values = np.array(value)

            bincounts, bin_edges = np.histogram(values)

            hist = summary_pb2.HistogramProto()
            hist.min = float(np.min(values))
            hist.max = float(np.max(values))
            hist.num = int(np.prod(values.shape))
            hist.sum = float(np.sum(values))
            hist.sum_squares = float(np.sum(values**2))

        # Add bin edges and counts
        for edge in bin_edges[1:]:
            hist.bucket_limit.append(edge)
        for v in bincounts:
            hist.bucket.append(v)

        summary = summary_pb2.Summary()
        summary.value.add(tag=tf_name, histo=hist)
        return summary

    def _scalar_summary(self, tf_name, value, step=None):
        summary = summary_pb2.Summary()
        summary.value.add(tag=tf_name, simple_value=value)
        return summary

    def _make_tf_name(self, name):
        tf_base_name = tf_name = make_valid_tf_name(name)
        i = 1
        while tf_name in self._tf_names:
            tf_name = '{}/{}'.format(tf_base_name, i)
            i += 1
        self._tf_names.add(tf_name)
        return tf_name

    def _log_summary(self, tf_name, summary, value, step=None):
        event = event_pb2.Event(wall_time=self._time(), summary=summary)
        if step is not None:
            event.step = int(step)
        if self.is_dummy:
            self.dummy_log[tf_name].append((step, value))
        else:
            self._write_event(event)

    def _write_event(self, event):
        data = event.SerializeToString()
        # See RecordWriter::WriteRecord from record_writer.cc
        w = self._writer.write
        header = struct.pack('Q', len(data))
        w(header)
        w(struct.pack('I', masked_crc32c(header)))
        w(data)
        w(struct.pack('I', masked_crc32c(data)))
        self._writer.flush()

    def _time(self):
        return self._dummy_time or time.time()

    def __del__(self):
        if self._writer is not None:
            self._writer.close()


def masked_crc32c(data):
    x = u32(crc32c(data))
    return u32(((x >> 15) | u32(x << 17)) + 0xa282ead8)


def u32(x):
    return x & 0xffffffff


def make_valid_tf_name(name):
    if not _VALID_OP_NAME_START.match(name):
        # Must make it valid somehow, but don't want to remove stuff
        name = '.' + name
    return '_'.join(_VALID_OP_NAME_PART.findall(name))


_default_logger = None  # type: Logger


def configure(logdir, flush_secs=2):
    """ Configure logging: a file will be written to logdir, and flushed
    every flush_secs.
    """
    global _default_logger
    if _default_logger is not None:
        raise ValueError('default logger already configured')
    _default_logger = Logger(logdir, flush_secs=flush_secs)


def _check_default_logger():
    if _default_logger is None:
        raise ValueError(
            'default logger is not configured. '
            'Call tensorboard_logger.configure(logdir), '
            'or use tensorboard_logger.Logger')


def log_value(name, value, step=None):
    _check_default_logger()
    _default_logger.log_value(name, value, step=step)


def log_histogram(name, value, step=None):
    _check_default_logger()
    _default_logger.log_histogram(name, value, step=step)


def log_images(name, images, step=None):
    _check_default_logger()
    _default_logger.log_images(name, images, step=step)

log_value.__doc__ = Logger.log_value.__doc__
